crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -540,6 +528,23 @@ where
 
         Ok(all_logs)
     }
+
+    /// Retrieves receipts and block from cache if near the tip (4 blocks), otherwise only receipts.
+    async fn receipts_and_maybe_block(
+        &self,
+        block_num_hash: &BlockNumHash,
+        best_number: u64,
+    ) -> Result<Option<(Arc<Vec<Receipt>>, Option<SealedBlock>)>, EthFilterError> {
+        let receipts_block = if (best_number - 4..=best_number).contains(&block_num_hash.number) {
+            self.eth_cache
+                .get_block_and_receipts(block_num_hash.hash)
+                .await?
+                .map(|(b, r)| (r, Some(b)))
+        } else {
+            self.eth_cache.get_receipts(block_num_hash.hash).await?.map(|r| (r, None))
+        };
+        Ok(receipts_block)
+    }
```

Comment:
we can even make this a function of the cache itself, that only fetches the receipts, and returns None for the block if not cached.

but can do that in a good first issue